src/System.Net.Sockets/tests/FunctionalTests/UnixDomainSocketTest.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+using System.Net.Test.Common;
+using System.Text;
+using System.Threading;
+
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.Net.Sockets.Tests
+{
+    public class UnixDomainSocketTest
+    {
+        private readonly ITestOutputHelper _log;
+
+        public UnixDomainSocketTest(ITestOutputHelper output)
+        {
+            _log = TestLogging.GetInstance();
+        }
+
+        private void OnConnectAsyncCompleted(object sender, SocketAsyncEventArgs args)
+        {
+            ManualResetEvent complete = (ManualResetEvent)args.UserToken;
+            complete.Set();
+        }
+
+        [Fact]
+        [PlatformSpecific(PlatformID.Windows)]
+        public void Socket_CreateUnixDomainSocket_Throws_OnWindows()
+        {
+            // Throws SocketException with this message "An address incompatible with the requested protocol was used"
+            Assert.Throws<SocketException>(() => new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified));
+        }
+
+        [Fact]
+        [PlatformSpecific(PlatformID.Linux | PlatformID.OSX)]
+        public void Socket_ConnectAsyncUnixDomainSocketEndPoint_Success()
+        {
+            string path = null;
+            SocketTestServer server = null;
+            UnixDomainSocketEndPoint endPoint = null;
+
+            for (int attempt = 0; attempt < 5; attempt++)
+            {
+                path = GetRandomNonExistingFilePath();
+                endPoint = new UnixDomainSocketEndPoint(path);
+                try
+                {
+                    server = SocketTestServer.SocketTestServerFactory(endPoint, ProtocolType.Unspecified);
+                    break;
+                }
+                catch (SocketException)
+                {
+                    // Path selection is contingent on a successful Bind(). 
+                    // If it fails, the next iteration will try another path.
+                }
+            }
+
+            Assert.NotNull(server);
+
+            SocketAsyncEventArgs args = new SocketAsyncEventArgs();
+            args.RemoteEndPoint = endPoint;
+            args.Completed += OnConnectAsyncCompleted;
+
+            ManualResetEvent complete = new ManualResetEvent(false);
+            args.UserToken = complete;
+
+            Socket sock = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
+            Assert.True(sock.ConnectAsync(args));
+
+            complete.WaitOne();
+
+            Assert.Equal(SocketError.Success, args.SocketError);
+            Assert.Null(args.ConnectByNameError);
+
+            complete.Dispose();
+            sock.Dispose();
+            server.Dispose();
+
+            File.Delete(path);
```

Comment:
Nit: we might want this to be in a `finally` block to ensure that we run cleanup even if the test fails.